### PR TITLE
Backfilling tests of enketo-webform.js

### PR DESCRIPTION
**Pull request by eyelidlessness**
### Background

The `enketo-webform.js` module is Enketo's main app entrypoint. On page load, it performs a series of initialization steps to bootstrap a form, for both online and offline modes. Many of these steps are order-sensitive, gradually building up a `Survey` representation of the form from various online/offline resources.

We've had several regressions since #269 related to changes here (as well as a few similar modules). The intent of this PR is to add test coverage to reduce the likelihood of future regressions.

### Some caveats

- While there's been much effort to avoid making changes to untested modules in the course of backfilling these tests, some minor changes were made to `enketo-webform.js` itself, and `translator.js`, to allow tests to initiate their behavior. Prior to those changes, their behavior is performed automatically. I think those changes are relatively low-risk, but wanted to call them out nonetheless.
- Quite a lot of functionality is not publicly exported, from those modules or their dependencies. With some caution around unintentionally expanding public APIs, a test-only build plugin was added to identify and export private top-level functions and variables from each app module. This depends on static analysis of the AST, and happens automatically in test. This may be undesirable, and may actually be irrelevant depending on the outcome of #300. (Worth noting that this build plugin is tested as well!)
- Those test-only private exports generally cannot be mocked directly, as the mocking depends on object property access rather than direct local references. They're only useful for calling directly within tests.
- Some of the current mocks are less than ideal, depending on specific implementation details. As test coverage improves, refactoring should be safer to improve test resiliency further as well.
- There's currently a fair bit of test helper code. I'm usually hesitant to use test helpers, preferring clarity over concision, but in this case there are several very similar steps repeated for both online/offline modes and I felt it easier to write and read the tests with these helpers. I'm certainly open to disagreement on that.

### To do

- [x] Test initialization steps performed, and order executed, for offline mode
- [x] Test initialization steps performed, and order executed, for online mode
- [x] Test failure of initialization steps
- [x] Underlying behavior and relevant branching logic of those steps
- [x] Minor refactors to fix some issues discovered in the course of this PR
    - [x] Unexpected/unguarded reference to undefined DOM nodes
    - [x] Readability improvements to `_prepareInstance`
- [x] Validate tests against changes in #293 (after removing workarounds). See https://github.com/eyelidlessness/enketo-express/commit/c975f14045aac058330984ba2c5f5db23b896462
- [x] Determine whether those steps are actually at a good level of abstraction for these tests to be useful for preventing regressions. The changes in #293 required minimal changes in https://github.com/eyelidlessness/enketo-express/commit/c975f14045aac058330984ba2c5f5db23b896462:
    - Removal of form-cache workaround steps
    - Adjustment for actual `controller.init` interface/call site changes
    - [Passes in CI](https://app.travis-ci.com/github/eyelidlessness/enketo-express/builds/236033821)

### CODE REVIEW

File: test/client/enketo-webform.spec.js [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+ * @property {Stub<[], Promise<void>} update
+ */
+
+/**
+ * @typedef ServiceWorkerStub
+ * @template State
+ * @property {ServiceWorkerControllerStub} [controller]
+ * @property {Stub<[ string ], Promise<ServiceWorkerRegistrationStub>>} register
+ * @property {State} state
+ */
```
Comment (eyelidlessness) [inline test/client/enketo-webform.spec.js:42]: At time of writing: some of these types were mistakenly left in from an earlier approach, but may still be relevant for tests that are as yet unwritten. Just want to note that this is likely noise at the moment.

**Comment by eyelidlessness** [comment]
@seadowg Apologies for comment/notification noise. GitHub isn't letting me request your review, I'm hoping the mention will fix that.

File: test/client/enketo-webform.spec.js [hunk 2]
```diff
@@ -0,0 +1,1731 @@
...
+                        <div class="form-header"></div>
+                    </div>
+                </div>
+                <div class="main-loader"></div>
+            `, 'text/html' );
+
+            mainElement = formDOM.documentElement.querySelector( '.main' );
+            loaderElement = formDOM.documentElement.querySelector( '.main-loader' );
+
+            document.body.append( mainElement, loaderElement );
```
Comment (eyelidlessness) [inline test/client/enketo-webform.spec.js:65]: These are accessed by enketo-webform at the top level on import/load, to define variables used by other functions during initialization. While those could also be wrapped in a function to provide for more deliberate behavior in test, this had to be done first before I felt changes like that would be safe. I'm up to change that now if preferred.

File: test/client/enketo-webform.spec.js [hunk 3]
```diff
@@ -0,0 +1,1731 @@
...
+         * `options.expectedArgs` items may be:
+         *
+         * - an instance of `ParameterPredicate`: its predicate will be performed
+         *   against the corresponding argument.
+         * - any other value: will be compared for deep equality.
+         *
+         * @param {InitStepOptions} options
+         * @return {InitStep}
+         */
+        const prepareInitStep = ( options ) => {
```
Comment (eyelidlessness) [inline test/client/enketo-webform.spec.js:264]: As mentioned in the PR notes, I tend to prefer not writing test helpers of this sort. They tend to grow options and branchy logic over time, and tend to make tests harder to understand and maintain. In this case there's so much boilerplate involved that it felt worth the risk for now. I'm hoping the named `options` and the presence of a `description` property is helpful enough for at least making sure the usage is understandable.

Another risk is that backfilling tests for enketo-webform-edit.js and enketo-webform-view.js will likely have similar mocking boilerplate needs. There would be a temptation to extract this helper and related functionality to some shared module, further removing test logic from where the tests are.

My inclination, to mitigate these risks:

- When I move on to backfill those tests, test *all* of those modules in one file, to keep the helper as close to the tests as possible.
- Gradually move toward more directly testable initialization logic to eliminate the need for this helper altogether.

**Review by eyelidlessness — COMMENTED** [review]
Added a handful of notes inline to hopefully add some helpful context for reviewers. I'm hoping the rest of the tests are relatively self-explanatory without additional commentary, but of course I'm happy to clarify wherever needed!

File: public/js/src/enketo-webform.js [hunk 4]
```diff
@@ -63,21 +95,41 @@ function _updateMaxSizeSetting( survey ) {
...
-    loader.classList.add( 'fail' );
+    loader.classList.add( LOAD_ERROR_CLASS );
 
     if ( error.status === 401 ) {
-        window.location.href = `${settings.loginUrl}?return_url=${encodeURIComponent( window.location.href )}`;
+        _location.href = `${settings.loginUrl}?return_url=${encodeURIComponent( _location.href )}`;
     } else {
         if ( !Array.isArray( error ) ) {
-            error = [ error.message  || t( 'error.unknown' ) ];
+            error = [ error.message  || t( LOAD_ERROR_UNKNOWN ) ];
```
Comment (MartijnR) [inline public/js/src/enketo-webform.js:123]: Just a single random comment so I won't forget. 

We used to have a build task called `grunt shell: translation` that updated the translation keys in locales/source/en/translation.json. It broke at some point and was disabled but I found it useful and wanted to bring it back eventually. https://github.com/i18next/i18next-parser

This parser doesn't support variables only string literals (as `t()` parameters). 

What do you think of the benefits of that (and manually updating translation keys) vs the benefits of exporting these for tests?

File: public/js/src/enketo-webform.js [hunk 5]
```diff
@@ -127,36 +188,42 @@ function _setEmergencyHandlers() {
...
-    if ( brandImg && survey.branding && survey.branding.source && brandImg.src !== survey.branding.source ) {
-        brandImg.src = '';
-        brandImg.setAttribute( attribute, survey.branding.source );
+    if ( brandImg != null ) {
+        if ( survey.branding && survey.branding.source && brandImg.src !== survey.branding.source ) {
+            brandImg.removeAttribute( 'src' );
+            brandImg.setAttribute( attribute, survey.branding.source );
+        }
+
+        brandImg.classList.remove( 'hide' );
```
[the remark this answers is not in the document]
Comment (MartijnR) [reply public/js/src/enketo-webform.js:216]: `.form-header__branding  img` is not always available (see _form-header.pug).

File: public/js/src/enketo-webform.js (hunk 4 above)
Comment (eyelidlessness) [reply public/js/src/enketo-webform.js:123]: Thanks for catching this!

I'm not sure I have an opinion either way about how translation keys are updated, I haven't really touched that at all so I'm not familiar. I didn't realize this depended on a parser at all. Is that necessary? It seems like it's doing the equivalent of `lodash.get`.

That said, addressing that here definitely feels out of scope, so my inclination is to revert to string literals and use the same string literals in the tests.

**Review by seadowg — CHANGES_REQUESTED** [review]
We discussed these changes on a call, but sticking them here as well:

* Describe(“initialization behaviour”) => describe(“helper functions”)
* Typedefs nearer use
* Multiple files
* Inline `prepareInitStep` (with sinon verification)
* Comment about private test

I spent the last twenty minutes randomly changing code in `enketo-webform.js` to see if I could find a broken/green tests state, but couldn't, which bodes well for coverage here. One thing I noticed in addition to the above notes is that the logging during tests is pretty verbose. Personally I think it would just make sense to remove logging as I'm not sure what value it adds.

**Comment by eyelidlessness** [comment]
> I spent the last twenty minutes randomly changing code in enketo-webform.js to see if I could find a broken/green tests state, but couldn't, which bodes well for coverage here.

This is great to hear! Still hoping to get coverage restored, but this is definitely encouraging.

> One thing I noticed in addition to the above notes is that the logging during tests is pretty verbose. Personally I think it would just make sense to remove logging as I'm not sure what value it adds.

Yeah I mostly had that in for debugging as I worked on this (believe it or not logger debugging can be less painful than breakpoints for some stuff! Especially scenarios where you’re likely to hit multiple breakpoints for the same reason). I wasn’t sure if it still provided value, so I threw it behind a flag. But if it’s noisy/distracting I’m happy to just remove it.

**Comment by eyelidlessness** [comment]
@seadowg I'm beginning to work on #366. We had previously discussed:

1. Merging in master to show test breakage from changes since these tests were written.
2. Rebasing to clean up history.
3. Merging as-is rather than addressing the changes you requested in initial review, with the expectation that these tests will either be removed or pared down and cleaned up in the course of the refactor.

Do you still agree that's the best next step? If so, I'd like to get this merged so I can open the next PR for similar test backfills of enketo-webform-edit.js and enketo-webform-view.js.

**Comment by seadowg** [comment]
@eyelidlessness sounds good! I'll try and take a look when I can but happy for you to go ahead with the rebase if I'm blocking you.

**Comment by eyelidlessness** [comment]
Closing in favor of a different approach, as [discussed here](https://github.com/enketo/enketo-express/issues/366#issuecomment-1015854097),

**Pull request closed without merging**